tests/unit/test_p11wrapper.cpp

```diff
@@ -0,0 +1,403 @@
+            .WillOnce(Return(0));
+
+    EXPECT_NO_THROW(_PKCS11_enumerate_slots(::testutils::somePKCS11CtxPtr()));
+}
+
+/* Test _PKCS11_release_all_slots() with NULL context. */
+TEST_F(P11WrapperTest, badSlotRelease)
+{
+    P11_SlotInfo slotInfo(::testutils::somePKCS11SlotPtr(), 2);
+    EXPECT_THROW(_PKCS11_release_all_slots(nullptr, slotInfo), P11Exception);
+}
+
+/* Test _PKCS11_release_all_slots() with bad slot information. */
+TEST_F(P11WrapperTest, badSlotRelease2)
+{
+    P11_SlotInfo slotInfo;
+    EXPECT_THROW(_PKCS11_release_all_slots(::testutils::somePKCS11CtxPtr(), slotInfo),
+                 P11Exception);
+}
+
+/* Test successful _PKCS11_release_all_slots().
+ */
+TEST_F(P11WrapperTest, goodSlotRelease)
+{
+    P11_SlotInfo slotInfo(::testutils::somePKCS11SlotPtr(), 2);
+
+    EXPECT_CALL(_mock(),
+                P11_PKCS11_release_all_slots(
+                        ::testutils::somePKCS11CtxPtr(), slotInfo._slots, slotInfo._numSlots))
+            .WillOnce(Return());
+
+    EXPECT_NO_THROW(_PKCS11_release_all_slots(::testutils::somePKCS11CtxPtr(), slotInfo));
+}
+
+/* Test _PKCS11_find_token() with NULL PKCS11 context. */
+TEST_F(P11WrapperTest, badSlotFinding)
+{
+    P11_SlotInfo slotInfo(::testutils::somePKCS11SlotPtr(), 2);
+    EXPECT_THROW(_PKCS11_find_token(nullptr, slotInfo), P11Exception);
+}
+
+/* Test P11_PKCS11_find_token() with bad value. */
+TEST_F(P11WrapperTest, badSlotFinding2)
+{
+    P11_SlotInfo slotInfo(::testutils::somePKCS11SlotPtr(), 2);
+
+    EXPECT_CALL(_mock(),
+                P11_PKCS11_find_token(
+                        ::testutils::somePKCS11CtxPtr(), slotInfo._slots, slotInfo._numSlots))
+            .WillOnce(Return(nullptr));
+
+    EXPECT_THROW(_PKCS11_find_token(::testutils::somePKCS11CtxPtr(), slotInfo), P11Exception);
+}
+
+/* Test successful P11_PKCS11_find_token(). */
+TEST_F(P11WrapperTest, goodSlotFinding)
+{
+    P11_SlotInfo slotInfo(::testutils::somePKCS11SlotPtr(), 2);
+
+    EXPECT_CALL(_mock(),
+                P11_PKCS11_find_token(
+                        ::testutils::somePKCS11CtxPtr(), slotInfo._slots, slotInfo._numSlots))
+            .WillOnce(Return(::testutils::somePKCS11SlotPtr()));
+
+    EXPECT_NO_THROW(_PKCS11_find_token(::testutils::somePKCS11CtxPtr(), slotInfo));
+}
+
+/* Test _PKCS11_open_session() with NULL slot pointer. */
+TEST_F(P11WrapperTest, badOpenSession)
+{
+    EXPECT_THROW(_PKCS11_open_session(nullptr, SessionMode::ReadWrite), P11Exception);
+}
+
+/* Test _PKCS11_open_session() with error return value. */
+TEST_F(P11WrapperTest, badOpenSession2)
+{
+    EXPECT_CALL(_mock(), P11_PKCS11_open_session(::testutils::somePKCS11SlotPtr(), 1))
+            .WillOnce(Return(-1));
+
+    EXPECT_THROW(_PKCS11_open_session(::testutils::somePKCS11SlotPtr(), SessionMode::ReadWrite),
+                 P11Exception);
+}
+
+/* Test successful _PKCS11_open_session(). */
+TEST_F(P11WrapperTest, goodOpenSession)
```

Comment:
Please use parameters and avoid code duplication.

Reply:
Done.